source/custom/custom.go

```diff
@@ -90,58 +98,35 @@ func (s Source) Discover() (source.Features, error) {
 // Process a single feature by Matching on the defined rules.
 // A feature is present if all defined Rules in a MatchRule return a match.
 func (s Source) discoverFeature(feature FeatureSpec) (bool, error) {
-	for _, rule := range feature.MatchOn {
-		// PCI ID rule
-		if rule.PciID != nil {
-			match, err := rule.PciID.Match()
-			if err != nil {
-				return false, err
-			}
-			if !match {
-				continue
-			}
-		}
-		// USB ID rule
-		if rule.UsbID != nil {
-			match, err := rule.UsbID.Match()
-			if err != nil {
-				return false, err
-			}
-			if !match {
-				continue
-			}
+	for _, matchRules := range feature.MatchOn {
+
+		allRules := []rules.Rule{
+			matchRules.PciID,
+			matchRules.UsbID,
+			matchRules.LoadedKMod,
+			matchRules.CpuID,
+			matchRules.Kconfig,
+			matchRules.Nodename,
 		}
-		// Loaded kernel module rule
-		if rule.LoadedKMod != nil {
-			match, err := rule.LoadedKMod.Match()
-			if err != nil {
-				return false, err
-			}
-			if !match {
+
+		ruleFailed := false
+		for _, rule := range allRules {
+			if reflect.ValueOf(rule).IsNil() {
 				continue
 			}
-		}
-		// cpuid rule
-		if rule.CpuID != nil {
-			match, err := rule.CpuID.Match()
-			if err != nil {
+			if match, err := rule.Match(); err != nil {
 				return false, err
-			}
-			if !match {
-				continue
+			} else if !match {
+				ruleFailed = true
```

Comment:
not sure if its favored but you could use `continue <label>` statement to continue to the next `matchRule` on the outer loop.

You would not need the `ruleFailed` bits in this case.
(just a suggestion here)

Reply:
yea no worries, im fine either way.

Reply:
> Personally I dislike those labels...

I'm not a big fan either, even if I've used them one or two times. You could achieve the above without labels e.g. by moving the inner loop into a separate function. But I think it's fine the way it is now and leave this to you to decide

Reply:
I like the idea of moving the inner loop into a func, will do that, thanks